digicampipe/scripts/pulse_shape.py

```diff
@@ -3,63 +3,211 @@
 Reconstruct the pulse template
 
 Usage:
-  pulse_shape.py [options] [FILE] [INPUT ...]
+  digicam-template [options] <outfile_path> <input_files>...
 
 Options:
   -h --help               Show this screen.
-  --max_events=N          Maximum number of events to analyse
-  -o FILE --output=FILE.  Output file.
-  -i INPUT --input=INPUT. Input files.
-  -c --compute            Compute the data.
-  -f --fit                Fit.
-  -d --display            Display.
-  -v --debug              Enter the debug mode.
-  -p --pixel=<PIXEL>      Give a list of pixel IDs.
+  --plots=PATH            path where to save a pdf full of interesting images.
+  --chunk_size INT        how many events to read before histogramming.
+                          This uses about 1GB memory, if you have less mem,
+                          reduce this number (will make the program slower).
+                          [default: 10000]
 '''
-import os
-from docopt import docopt
-from tqdm import tqdm
-from schema import Schema, And, Or, Use, SchemaError
 import numpy as np
-import matplotlib.pyplot as plt
-
+import h5py
 from digicampipe.io.event_stream import calibration_event_stream
-from digicampipe.io.containers_calib import CalibrationContainer
+
+from tqdm import tqdm
+from docopt import docopt
+import matplotlib.pyplot as plt
+from matplotlib.backends.backend_pdf import PdfPages
+
+from ctapipe.visualization.mpl import CameraDisplay
+from digicampipe.utils import DigiCam
+
+
+def fill_histo_from_buffer(histo, buffer):
+    for pid in tqdm(range(histo.shape[0])):
+        for sid in range(histo.shape[1]):
+            histo[pid, sid] += np.bincount(
+                buffer[:, pid, sid], minlength=histo.shape[2]
+            ).astype(histo.dtype)
+
+
+def main(outfile_path, input_files=[], chunk_size=10000, plots_path=None):
+    events = calibration_event_stream(input_files)
+
+    histo = None
+    buffer = None
+
+    for i, e in enumerate(tqdm(events)):
+        j = i % chunk_size
+        adc = e.data.adc_samples
+
+        if histo is None:
+            histo = np.zeros((*adc.shape, 4096), dtype='u2')
+
+        if j == 0:
+            if buffer is not None:
+                fill_histo_from_buffer(histo, buffer)
+            buffer = np.zeros((chunk_size, *adc.shape), dtype=adc.dtype)
+        buffer[j] = adc[...]
+    buffer = buffer[:j]
+    fill_histo_from_buffer(histo, buffer)
+
+    # store buffer zipped to a h5 file
+    outfile = h5py.File(outfile_path)
+    outfile.create_dataset(
+        name='adc_count_histo',
+        data=histo,
+        compression="gzip",
+        chunks=(4, histo.shape[1], 128)
+    )
+
+    mean = np.zeros(adc.shape, dtype='f4')
+    std = np.zeros(adc.shape, dtype='f4')
+
+    for pid in range(1296):
```

Comment:
let's avoid magic numbers here: we could use n_pixels = adc.shape[0] or len(adc). Also same comment as before in principle one can use the `histogram` package

Reply:
I think a README would help me understanding how to use hat package. 

I tried to understand it from the code, but as it is using C/C++ stuff I was mainly lost after 10 lines...